pandas/tests/indexes/test_setops.py

```diff
@@ -95,3 +103,38 @@ def test_union_dtypes(left, right, expected):
     b = pd.Index([], dtype=right)
     result = (a | b).dtype
     assert result == expected
+
+
+def test_union_duplicate_index_subsets_of_each_other():
+    # GH: 31326
+    a = pd.Index([1, 2, 2, 3])
+    b = pd.Index([3, 3, 4])
+    expected = pd.Index([1, 2, 2, 3, 3, 4])
+
+    result = a.union(b)
+    tm.assert_index_equal(result, expected)
+    result = a.union(b, sort=False)
+    tm.assert_index_equal(result, expected)
+
+
+@pytest.mark.parametrize(
+    "func",
+    [
+        (Int64Index),
+        (Float64Index),
+        (DatetimeIndex),
+        (CategoricalIndex),
+        (TimedeltaIndex),
```

Comment:
you dont needs parens around each of these

Reply:
can you include `lambda x: Index(x, dtype=object)`

Reply:
Done